chapters/part3/hyperparameter-optimization-with-optuna.md

```diff
@@ -0,0 +1,426 @@
+
+Additionally, you can export a configuration with optimized hyperparameters.
+
+```python
+dump_best_config("./imdb_optuna.jsonnet", "./best_config.json", study)
+```
+
+It will create a configuration named `best_config.json`.
+This is helpful to retrain a model with the best hyperparameters.
+
+You can try the example on [colab](https://colab.research.google.com/github/himkt/optuna-allennlp/blob/master/notebook/Optuna_AllenNLP.ipynb).
+
+</exercise>
+
+<exercise id="4" title="Writing your own script with Optuna and Use Pruners">
+
+Additionally, you can use Optuna when you define the AllenNLP model by writing your script.
+A python code of training a model with AllenNLP typically consists of main three parts: preparing data, building model, and creating trainer.
+In this section, we show the example of AllenNLP+Optuna with writing Python script.
+<!-- We put the python code to use AllenNLP+Optuna at the bottom of this exercise. -->
+
+## Pruning unpromising trials
+
+Hyperparameter optimization often takes a long time to find good hyperparameters.
+If you can find and stop unpromising trials with bad hyperparameters, you can reduce the time of optimization and get good hyperparameters faster.
+Stopping unpromising trials is so-called `pruning`.
+The following illustration shows an example of pruning.
+Although the final curves at the left side of the picture are not available before finishing hyperparameter optimization,
+a pruner evaluates at each epoch how promising each trial will be and stops if it predicts to be unpromising.
+
+<img src="/part3/hyperparameter-optimization-with-optuna/illustration_of_pruning.jpg" alt="Illustration of Pruning" />
+
+You can use `AllenNLPPruningCallback` that is the new feature of Optuna, which allows users to prune unpromising trials with algorithms implemented in Optuna.
+`AllenNLPPruningCallback` is the interface to provide the way to use these pruning algorithms in `GradientDescentTrainer`, which is the standard way to train a model in AllenNLP.
+To enable a pruning, you must create the AllenNLP callback first. `AllenNLPPruningCallback` needs two arguments: a trial and a target metric.
+In this example, we use `validation_accuracy` for the metric to determine if a trial should be pruned or not.
+Note that `patience` is set to `None` since pruning and built-in early stopping could conflict.
+The objective function looks like following:
+
+```python
+    def objective(trial):
+        trainer = GradientDescentTrainer(
+            model=model,
+            optimizer=optimizer,
+            data_loader=data_loader,
+            validation_data_loader=validation_data_loader,
+            validation_metric="+accuracy",
+            patience=None,
+            num_epochs=50,
+            cuda_device=CUDA_DEVICE,
+            serialization_dir="result",
+            epoch_callbacks=[AllenNLPPruningCallback(trial, "validation_accuracy")],
```

Comment:
This is interesting, I didn't know about this.  Is there a way to get the callback instantiated through a config file?  What would have to happen to make that work?  It's pretty annoying to have to instantiate the trainer yourself here, because it messes with a bunch of our typical recommendations for allennlp usage.  If we can make this easier, that would be better.

Reply:
We have a mechanism for passing in arbitrary, in-memory objects when using `FromParams` (see [here](https://guide.allennlp.org/using-config-files#4)).  The main problem is that you need a line somewhere in your code that passes in a `trial` object when you call `from_params`, and you need that value to be passed down to all of your objects recursively.  I tried to figure out a way to do that with minimal modification to our train command, and there isn't a super easy way.  But, here's a bit of a hack that you can do in your own code that abuses the `FromParams` code a bit, but I think it might actually work.

You have [a method](https://github.com/optuna/optuna/blob/9b0081a212cba04e97a28ca510e5e72559b01a0d/optuna/integration/allennlp.py#L144-L145) where you're instantiating a `Params` object.  If that object had a `trial` key with the `Trial` object as a value, it will get passed to `TrainModel.from_params`.  The trouble is that it won't be treated as an `extra`, which it needs to be in order to be passed down recursively to all child classes (like the callbacks that you want to construct).  So, what you can do is create your own subclass of `TrainModel`, registered with a separate name, and have that hijack the `from_params` method and turn the `trial` key from something in the `Params` dictionary to an `extra`.  It would look something like this:

```python
@TrainModel.register("optuna")
class TrainModelWithOptuna(TrainModel):
    def from_params(# match the arguments to `FromParams.from_params`):
        # This is a pretty crazy hack of how FromParams was
        # intended to be used, but it should work for your use case
        trial = params.pop("trial")
        return TrainModel.from_params(params=params, trial=trial, **extras)
```

When objects are passed as extra `kwargs` like this, they are made available to the constructors of every object that gets constructed underneath this object in the code, regardless of whether intermediate classes need that argument (this is how the `Embedding` has access to the `Vocabulary` inside of a `TextFieldEmbedder`, even though the `TextFieldEmbedder` doesn't take a `Vocabulary` as an argument).  So, if we can get `TrainModel.from_params` to get called with `trial=trial`, then the callbacks that you have should get passed the `trial` object correctly.

In order to make this work, where you [instantiate the `Params` object](https://github.com/optuna/optuna/blob/9b0081a212cba04e97a28ca510e5e72559b01a0d/optuna/integration/allennlp.py#L144-L145) you need to add a `"type": "optuna"` key, and a `"trial": trial` key.  The `FromParams` code will call `TrainModel.from_params` with your hacked parameter object, and it will see the type is `optuna`, and call `TrainModelWithOptuna.from_params` after _popping the `type` key_.  Then inside `TrainModelWithOptuna.from_params`, we pop the `trial` key and make it an `extra` when we call `TrainModel.from_params` again.  This time, there is no `type` key, so the default `TrainModel` constructor gets called, but it now has your extra argument.

That should be all that's necessary - only one extra class and a couple of lines of modification to your `run` logic.  Understanding why this works is complicated, though, so I'd definitely recommend having lots of comments in your code explaining what's going on, if you choose to do this.

Reply:
Sorry for the late reply.
I found the way that enables `AllenNLPExecutor` to use `AllenNLPPruningCallback`.

We are working on adding support for Optuna pruning API in https://github.com/optuna/optuna/pull/1772.
After merging this PR, we can easily enable pruning by specifying an epoch callback.
(e.g. https://github.com/himkt/optuna-allennlp/pull/9/files)

1cb7a7a  I updated the content based on the PR.